From 91bf51b055570a54266e65766aa434349f735652 Mon Sep 17 00:00:00 2001
From: Jakub Tyszkowski <jakub.tyszkowski@tieto.com>
Date: Fri, 14 Nov 2014 14:22:24 +0100
Subject: [PATCH 02/11] sepolicy: Add initial policy file for bluetoothd

This adds initial policies for selinux regarding bluetooth daemon.
It still not ready to use enforcing policies due to logwrapper issue
that needs to be handled.

Change-Id: Iaefdd7c7c5e7958074819cea0026945dd96f0b61
---
 bluetoothd.te | 40 ++++++++++++++++++++++++++++++++++++++++
 file_contexts |  6 ++++++
 2 files changed, 46 insertions(+)
 create mode 100644 bluetoothd.te

diff --git a/bluetoothd.te b/bluetoothd.te
new file mode 100644
index 0000000..f811b68
--- /dev/null
+++ b/bluetoothd.te
@@ -0,0 +1,40 @@
+type bluetoothd, domain;
+type bluetoothd_exec, exec_type, file_type;
+
+# FIXME: disabling this is the final goal
+permissive bluetoothd;
+
+# Start bluetoothd from init
+# allow init:bluetoothd_exec:file execute_no_trans;
+# its resolved to: allow init bluetoothd_exec:file { getattr open read execute };
+init_daemon_domain(bluetoothd)
+
+# FIXME: this does not seams to work for logwrapper
+domain_auto_trans(init, bluetoothd_exec, bluetoothd)
+# this is reported but neverallowed:
+# allow bluetoothd system_file:file entrypoint;
+
+# Data file accesses
+allow bluetoothd bluetooth_data_file:dir w_dir_perms;
+allow bluetoothd bluetooth_data_file:notdevfile_class_set { create_file_perms link_file_perms };
+
+# TODO: Verify if this can be narrowed
+allow bluetoothd self:capability { setuid net_admin net_bind_service };
+allow bluetoothd kernel:system module_request;
+
+# TODO: this may be romoved for userbuild where we don't use bluetoothd_wrapper
+allow bluetoothd bluetoothd_exec:file execute_no_trans;
+
+# TODO: minimize this set
+allow bluetoothd self:socket { create_socket_perms accept listen recv_msg send_msg lock recvfrom sendto name_bind relabelfrom relabelto };
+
+# stream sockets
+allow bluetoothd { bluetooth mediaserver }:unix_stream_socket connectto;
+allow bluetooth bluetoothd:unix_stream_socket rw_socket_perms;
+allow bluetooth bluetoothd:fd use;
+
+# needs /system/bin/log access
+allow bluetoothd devpts:chr_file rw_file_perms;
+
+# access to uhid device
+allow bluetoothd uhid_device:chr_file rw_file_perms;
diff --git a/file_contexts b/file_contexts
index d964f9b..131ce99 100644
--- a/file_contexts
+++ b/file_contexts
@@ -311,6 +311,12 @@
 /data/app-asec(/.*)?        u:object_r:asec_image_file:s0
 
 #############################
+# Bluetooth daemon
+/system/bin/bluetoothd             u:object_r:bluetoothd_exec:s0
+/system/bin/bluetoothd-main        u:object_r:bluetoothd_exec:s0
+/system/bin/bluetoothd-snoop       u:object_r:bluetoothd_exec:s0
+
+#############################
 # external storage
 /mnt/media_rw(/.*)?         u:object_r:mnt_media_rw_file:s0
 /mnt/user(/.*)?             u:object_r:mnt_user_file:s0
-- 
2.9.3

